### PR TITLE
Stats/BarGauge: allow setting text size

### LINKED ISSUE #23305: Enable font size option for Stat panel
I would like the possibility to change the font-size for the field title in the Stat panel. Right now I am showing the status about my hosts networkadapters where I have the networkadapter names in the font-title. 
But the text is so small I can bearly see it unless I resize the panel to take the whole screen.
I also searched the form and found several other questions about this but none had been answered.

### PR DESCRIPTION
**Pull request by ryantxu**
Fixes #23305  and helps #22209

Currently, most panels do not allow setting font sizes directly, rather the font sizes depend on the text we are trying to display.  We do this automatically for a couple reasons:
* allowing a fixed size does not scale well when the dashboard resizes
* the number of distinct font size options is potentially large.  For example, the bar gauge has different sizes for title, text, unit in both horizontal and vertical mode.
* as panel structures change, migrating the explicit settings gets really awkward

Under the hood, the automatic calculations are all based on the *longest* strings the visualization has to show.  In this PR, we let people configure the string lengths explicitly rather than calculating them.

|  Automatic | Explicit   | 
|---|---|
|  image | image |

I went though a few options to enter the text length, and am finally using a select -- any better ideas?  Typing the text felt awkward with slow refresh rates and unclear response

image

TODO:

- [ ] Docs for stats panel  
- [ ] Docs for bar gauge panel

(I'll look at the docs once we agree on a reasonable structure/UI)

### CODE REVIEW

File: public/app/features/dashboard/components/PanelEditor/PanelOptionsEditor.tsx [hunk 1]
```diff
@@ -57,11 +57,12 @@ export const PanelOptionsEditor: React.FC<PanelOptionsEditorProps<any>> = ({
               return null;
             }
 
-            const label = (
-              <Label description={e.description} category={e.category?.slice(1)}>
-                {e.name}
-              </Label>
-            );
+            const label =
```
Comment (ryantxu) [inline public/app/features/dashboard/components/PanelEditor/PanelOptionsEditor.tsx:60]: This change will only show the label if name|description exists.  This allows me to make the `<Label ` explicit in the body depending on the option state

File: public/app/plugins/panel/stat/AlignmentFactorText.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import { Input } from '@grafana/ui';
+import { css } from 'emotion';
+import { config } from 'app/core/config';
+
+interface Props {
+  value?: string;
+  onChange: (value?: string) => void;
+}
+
+export const AlignmentFactorTextEditor: React.FC<Props> = ({ value, onChange }) => {
```
Comment (ryantxu) [inline public/app/plugins/panel/stat/AlignmentFactorText.tsx:11]: NOTE:  this was my direct text option, and should be removed if we stick with the `<Select` version

**Comment by ryantxu** [comment]
@torkelo -- this is not great... but I am not sure what a better approach would be

File: public/app/plugins/panel/stat/AlignmentFactorLengthEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      label: getLabelText(i),
+    };
+  });
+
+export const AlignmentFactorLengthEditor: React.FC<Props> = ({ value, onChange }) => {
+  const styles = {
+    textInput: css`
+      margin-bottom: 8px;
+    `,
+  };
```
Comment (hugohaggmark) [inline public/app/plugins/panel/stat/AlignmentFactorLengthEditor.tsx:32]: nit: maybe move styles function outside of the component and use `useStyles` hook

File: public/app/plugins/panel/stat/AlignmentFactorLengthEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  const getTextLengthFrom = (txt: string) => {
+    if (!txt) {
+      return 0;
+    }
+    const len = parseInt(txt, 10);
+    if (isNaN(len)) {
+      return txt.length;
+    }
+    return len;
+  };
```
Comment (hugohaggmark) [inline public/app/plugins/panel/stat/AlignmentFactorLengthEditor.tsx:53]: maybe move this one outside of the component?

**Review by hugohaggmark — COMMENTED** [review]
I think the code looks great and I just left some minor nits. 

As a first time user of the UX I find the Explicit factors a bit complicated to understand. Can the 8 - xxxxx be visualized in another way? Maybe with a slider and showing in real time how the Font size is changed?

**Comment by ryantxu** [comment]
> Explicit factors a bit complicated to understand.

:rofl: no kidding

I'm not sure the best solution... in the end I think it may be exposing explicit font sizes!  but that is a bigger project.  uggg

**Comment by torkelo** [comment]
This was more complex than I imagined, do we need this many separate settings? I imaged it would be enough with title & value

**Comment by ryantxu** [comment]
I removed the 7.3 label since this feels too weird, and I think we should likely just expose some standard real text sizes (like text and title)

**Comment by ryantxu** [comment]
Closing this in favor of #29476

**Pull request closed without merging**